In [203]:
import pandas as pd

In [204]:
customers = pd.read_csv('customers.txt',sep=',')
customers.head()

CustomerID    FirstName      LastName         City State
0        6192        Randi    Piedrahita  Canoga Park    CA
1     5100595  Christopher       Abraham   Loma Linda    CA
2     1902451        Elvis      Jauregui  Los Angeles    CA
3     1906377         Jose  Mota-Sanchez       Downey    CA
4     1902406      Scott        McCrorie   Long Beach    CA

In [205]:
orders = pd.read_csv('orders.txt',sep=',')
orders.head()

CustomerID  OrderID                     Date  OrderTotal ProductName  Price
0        8091  7742581  2021-07-26 14:40:10.783        95.0         Z03   90.0
1      902139  7742778  2021-08-08 05:01:21.120        60.0    0844 A/C   60.0
2     2300266  7742593  2021-07-27 11:00:16.020       185.0         M07   90.0
3     2300266  7742593  2021-07-27 11:00:16.020       185.0         M09   90.0
4     5173013  7742609  2021-07-28 14:26:13.930       165.0        0324  160.0

## Part 1 (Only customers)

In [206]:
customers_by_id = customers['CustomerID'].drop_duplicates()
customers_by_id.count()

471

There is 471 diferents customers

In [207]:
states = customers['State'].drop_duplicates()
states
#15 pero hay un error

0                 CA
23                AZ
65                NV
76                FL
95     CA           
99                WA
108               NH
129               ID
164               CO
203               Ca
207               TX
242               NM
350               OR
499               UT
565               MA
578               IN
Name: State, dtype: object

In [208]:
result_state = customers.groupby('State')['CustomerID'].count()
result_state

State
AZ                 6
CA               562
CA                 5
CO                 3
Ca                 2
FL                 3
ID                 1
IN                 1
MA                 1
NH                 1
NM                 3
NV                 8
OR                 1
TX                 2
UT                 2
WA                 1
Name: CustomerID, dtype: int64

In [209]:
result_ln=customers.groupby('LastName')['CustomerID'].count().drop_duplicates()
result_ln 

LastName
Abraham    2
Abrams     1
Brady      3
Doggett    5
Garcia     4
Smith      8
Name: CustomerID, dtype: int64

## Part 2

In [210]:
orders_by_id = orders['OrderID'].drop_duplicates()
orders_by_id.count()

16672

There is 16672 diferents orders

In [211]:
result = orders.groupby('OrderID')['ProductName'].count()
result.mean()

1.7570777351247602

The average number of items per order is 1.75 or 1.76

In [212]:
result.max()

35

The highest number of items per order is 35

In [213]:
# Filter data between two dates
orders['Date'] = pd.to_datetime(orders['Date'], format='%Y-%m-%d')
filtered_orders = orders.loc[(orders['Date'] >= '2021-10-01')
                     & (orders['Date'] < '2021-11-01')]
filtered_orders['OrderID'].drop_duplicates().count()

267

The number of orders placed in October 2021 was 267

In [214]:
df=pd.DataFrame()
df['Month'] = pd.DatetimeIndex(orders['Date']).month
orders['Month']=df
df

Month
0        7.0
1        8.0
2        7.0
3        7.0
4        7.0
...      ...
29289    2.0
29290    2.0
29291    2.0
29292    2.0
29293    2.0

[29294 rows x 1 columns]

In [215]:
orders

CustomerID  OrderID                    Date  OrderTotal  ProductName  \
0            8091  7742581 2021-07-26 14:40:10.783        95.0          Z03   
1          902139  7742778 2021-08-08 05:01:21.120        60.0     0844 A/C   
2         2300266  7742593 2021-07-27 11:00:16.020       185.0          M07   
3         2300266  7742593 2021-07-27 11:00:16.020       185.0          M09   
4         5173013  7742609 2021-07-28 14:26:13.930       165.0         0324   
...           ...      ...                     ...         ...          ...   
29289     5415451  7791815 2022-02-07 14:37:30.567         0.0  Circle #041   
29290     5415483  7791820 2022-02-07 15:55:55.260       132.0         HH33   
29291     5415485  7791822 2022-02-07 16:23:05.940       132.0         3568   
29292     4300695  7791824 2022-02-07 16:55:33.343       157.0         0414   
29293     5415503  7791848 2022-02-09 10:36:56.553       132.0         PP22   

       Price  Month  
0       90.0    7.0  
1       60.0    8.0  
2       90.0    7.0  
3       90.0    7.0  
4      160.0    7.0  
...      ...    ...  
29289    0.0    2.0  
29290  125.0    2.0  
29291  125.0    2.0  
29292  150.0    2.0  
29293  125.0    2.0  

[29294 rows x 7 columns]

In [216]:
orders_not_null = orders.dropna()
orders_not_null

CustomerID  OrderID                    Date  OrderTotal  ProductName  \
0            8091  7742581 2021-07-26 14:40:10.783        95.0          Z03   
1          902139  7742778 2021-08-08 05:01:21.120        60.0     0844 A/C   
2         2300266  7742593 2021-07-27 11:00:16.020       185.0          M07   
3         2300266  7742593 2021-07-27 11:00:16.020       185.0          M09   
4         5173013  7742609 2021-07-28 14:26:13.930       165.0         0324   
...           ...      ...                     ...         ...          ...   
29289     5415451  7791815 2022-02-07 14:37:30.567         0.0  Circle #041   
29290     5415483  7791820 2022-02-07 15:55:55.260       132.0         HH33   
29291     5415485  7791822 2022-02-07 16:23:05.940       132.0         3568   
29292     4300695  7791824 2022-02-07 16:55:33.343       157.0         0414   
29293     5415503  7791848 2022-02-09 10:36:56.553       132.0         PP22   

       Price  Month  
0       90.0    7.0  
1       60.0    8.0  
2       90.0    7.0  
3       90.0    7.0  
4      160.0    7.0  
...      ...    ...  
29289    0.0    2.0  
29290  125.0    2.0  
29291  125.0    2.0  
29292  150.0    2.0  
29293  125.0    2.0  

[27152 rows x 7 columns]

In [217]:
best_month=orders_not_null.groupby('Month')['Price','Month'].head()
best_month.head()

C:\Users\Leo\AppData\Local\Temp\ipykernel_4668\1654385141.py:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  best_month=orders_not_null.groupby('Month')['Price','Month'].head()


Price  Month
0   90.0    7.0
1   60.0    8.0
2   90.0    7.0
3   90.0    7.0
4  160.0    7.0

In [218]:
print(best_month.groupby('Month')['Price'].sum().max())
best_month.groupby('Month')['Price'].sum()

600.0


Month
1.0     410.0
2.0     560.0
3.0     340.0
4.0     465.0
5.0     470.0
6.0     600.0
7.0     555.0
8.0     240.0
9.0     585.0
10.0    290.0
11.0    590.0
12.0    575.0
Name: Price, dtype: float64

The month with the highest profit was july